src/docker.rs

```diff
@@ -98,145 +193,412 @@ fn parse_docker_opts(value: &str) -> Result<Vec<String>> {
+        });
+
+        // root is either workspace_root, or, if we're outside the workspace root, the current directory
+        let mount_root: PathBuf;
+        #[cfg(target_os = "windows")]
+        {
+            // On Windows, we can not mount the directory name directly. Instead, we use wslpath to convert the path to a linux compatible path.
+            mount_root = wslpath(&host_root, verbose)?;
+        }
+        #[cfg(not(target_os = "windows"))]
+        {
+            mount_root = mount_finder.find_mount_path(host_root.clone());
+        }
+        let mount_cwd: PathBuf;
+        #[cfg(target_os = "windows")]
+        {
+            // On Windows, we can not mount the directory name directly. Instead, we use wslpath to convert the path to a linux compatible path.
+            mount_cwd = wslpath(cwd, verbose)?;
+        }
+        #[cfg(not(target_os = "windows"))]
+        {
+            mount_cwd = mount_finder.find_mount_path(cwd);
+        }
+        let sysroot = mount_finder.find_mount_path(sysroot);
+
+        Ok(Directories {
+            cargo,
+            xargo,
+            target,
+            nix_store,
+            host_root,
+            mount_root,
+            mount_cwd,
+            sysroot,
+        })
+    }
+}
+
+#[derive(Debug)]
+enum VolumeId {
+    Keep(String),
+    Discard(String),
+}
+
+impl VolumeId {
+    fn create(engine: &Engine, container: &str, verbose: bool) -> Result<Self> {
+        let keep_id = format!("{container}-keep");
+        if volume_exists(engine, &keep_id, verbose)? {
+            Ok(Self::Keep(keep_id))
+        } else {
+            Ok(Self::Discard(container.to_string()))
+        }
+    }
+}
+
+impl AsRef<str> for VolumeId {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::Keep(s) => s,
+            Self::Discard(s) => s,
+        }
+    }
+}
+
+fn cargo_cmd(uses_xargo: bool) -> SafeCommand {
+    if uses_xargo {
+        SafeCommand::new("xargo")
+    } else {
+        SafeCommand::new("cargo")
+    }
+}
+
 #[allow(unused_variables)]
-pub fn mount(cmd: &mut Command, val: &Path, verbose: bool) -> Result<PathBuf> {
-    let host_path = file::canonicalize(&val)
-        .wrap_err_with(|| format!("when canonicalizing path `{}`", val.display()))?;
-    let mount_path: PathBuf;
+fn canonicalize_mount_path(path: &Path, verbose: bool) -> Result<PathBuf> {
     #[cfg(target_os = "windows")]
     {
         // On Windows, we can not mount the directory name directly. Instead, we use wslpath to convert the path to a linux compatible path.
-        mount_path = wslpath(&host_path, verbose)?;
+        wslpath(path, verbose)
     }
     #[cfg(not(target_os = "windows"))]
     {
-        mount_path = host_path.clone();
+        Ok(path.to_path_buf())
     }
-    cmd.args(&[
+}
+
+fn canonicalize_host_path(path: &Path) -> Result<PathBuf> {
+    file::canonicalize(path).wrap_err_with(|| format!("when canonicalizing path `{path:?}`"))
+}
```

Comment:
why it's own function?

Reply:
I just broke it up because it's used in many places and wanted the same error message upon failure. I could move it easily into `file::canonicalize` though, that is a good idea.

Reply:
add the error message to `file::canonicalize`, good to have it standardized